features/nanostack/mbed-mesh-api/source/NanostackEMACInterface.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2017 ARM Limited. All rights reserved.
+ */
+
+#include "Nanostack.h"
+#include "NanostackEthernetInterface.h"
+#include "NanostackEthernetPhy.h"
+#include "EMAC.h"
+#include "nsdynmemLIB.h"
+#include "arm_hal_phy.h"
+
+#if 1
+class EMACPhy : public NanostackEthernetPhy
+{
+public:
+    EMACPhy(NanostackMemoryManager &mem, EMAC &m);
+    virtual int8_t phy_register();
+    virtual void get_mac_address(uint8_t *mac);
+    virtual void set_mac_address(uint8_t *mac);
+
+    int8_t address_write(phy_address_type_e , uint8_t *);
+    int8_t tx(uint8_t *data_ptr, uint16_t data_len, uint8_t tx_handle,data_protocol_e data_flow);
+
+    void emac_phy_rx(emac_mem_buf_t *mem);
+
+private:
+    NanostackMemoryManager &memory_manager;
+    EMAC &emac;
+    uint8_t mac_addr[6];
+    int8_t device_id;
+    phy_device_driver_s phy;
+};
+
+// GAH! no handles on the callback. Force a single interface
+static EMACPhy *single_phy;
+
+extern "C"
+{
+static int8_t emac_phy_address_write(phy_address_type_e address_type, uint8_t *address_ptr)
+{
+    return single_phy->address_write(address_type, address_ptr);
+
+}
+
+static int8_t emac_phy_interface_state_control(phy_interface_state_e, uint8_t)
+{
+    return -1;
+}
+
+static int8_t emac_phy_tx(uint8_t *data_ptr, uint16_t data_len, uint8_t tx_handle,data_protocol_e data_flow)
```

Comment:
The name of this function could be tx(), and the one exposed as an API should be emac_py_tx(). 
data_ptr should be const.

Reply:
Naming scheme appears to be that all the static functions we're providing as Nanostack callback functions are `emac_phy_`. Not sure I see a problem with that - I think they're consistent at least.

We can't change the prototype, alas - they're very historical Nanostack driver attachment points, used by third party drivers. Maybe worth attempting, but would need a new Nanostack release, and modifying all drivers, including this glue later.

And the public API including the 'tx' is inherited from `NanostackPhy` which is also public API being implemented by third parties